### ISSUE TITLE
GC optimization: don't scan immutable arrays of scalars.

### ISSUE THREAD

**Issue by crusso** [dfinity/motoko#3011]
The GC doesn't know if an array is mutable, immutable or immutable with scalar contents.

For an _immutable_ array, the first GC scan could detect if all entries are scalar and set a bit in the header to prevent future marking.

We can't do this for mutable arrays because some entries can change from scalar to reference type after an update (eg. Nat which have mixed scalar/boxed representations, perhaps also ).

Is this worth it? At least until we have packed array representations?

**Comment by crusso** [comment]
@ulan @ggreif

**Comment by nomeata** [comment]
Fun, I was just thinking of something similar a few days ago. But I thought about the following improvement: Create a new heap object type for arrays of scalar data (or add a flag to the heap tag, same thing.), and when generating code for allocation, if we statically know the value type is always a scalar (`Word8`, `Word16`, `Bool`), set that flag. Both for immutable and mutable.

Advantage: Not even a first GC pass needed; works also for mutable arrays.

Disadvantage: Wouldn’t kick in in polymorphic code. That probably kills it already.

Anyways, if you have it your way (GC-discoverable dynamic flag) we can easily later set that flag already

**Comment by nomeata** [comment]
Of course, the sensible thing to do would be to get a good instruction counting profile and see how much time is spent GC’ing arrays :-)

**Comment by ulan** [comment]
V8 uses something similar to @nomeata's idea. Arrays have element kinds which encode whether the array stores integers, doubles, pointers. So the idea works in practice.

I am not sure about the dynamic checks. We need to ensure that the overhead for detecting that all elements are scalar added during scanning doesn't exceed the benefit we get.

**Comment by crusso** [comment]
Or we do 1-bit type passing and use that to determine what to do. 

But if we go there, it might be better to pass a few more bits to encode the scalar+size (0,1,2,3,4) or reference and do packing/unpacking & GC based on that, assuming that's enough info to restore any scalar value to a vanilla rep.

**Comment by nomeata** [comment]
Related discussion about packed arrays for fixed-width data: https://github.com/dfinity/motoko/issues/966#issuecomment-573681851
